Do not load entire database for all commands

sruffell opened the pull request:
This merge request improves the performance when there are many entries in the database by not loading or parsing all the lines in many cases where it is not necessary.

Related to issue #245

sruffell commented:
I was just thinking about this, and I am thinking that it might make more sense for being/end to walk backwards through the inclusions. That would eliminate most of the use of rbegin/rend and put them in "natural" order since @1 is always the most recent, etc..

lauft commented:
You mean instead of adding an extra set of iterators (`rbegin`/`rend`), to redefine the existing ones? That's fine with me. As you already mentioned, the natural order for Timewarrior is going backwards in time through the intervals.

sruffell commented:
> You mean instead of adding an extra set of iterators (rbegin/rend), to redefine the existing ones?

The pull request already defines two sets of iterators begin/end and rbegin/rend.  I was just thinking that the semantics of the two sets should be reversed, which would allow more of code to use the ranged based for loops.

But ok, I'll update the pull request with that change.

sruffell commented:
I've updated this branch so that range based for loops on the database start with the most recent interval (and renamed the lastLine to firstLine to match).

Also as part of working to remove the users of getAllInclusions, I added an interface to the Database in order to get the set of tags directly from the tags database.  This leaves only a single user of getAllInclusions left.

sruffell commented:
Since the performance appears to be close to O(1) now, I do not intend to do any more work on this branch pending review from @lauft

sruffell commented:
I pushed another update. I realized that the original getIntervalsByIds would consider the latest interval twice if there were synthetic intervals, and this condition was not caught in any tests.

I updated two of the tests, annotate.t and move.t, to include a mix of non-synthetic and synthetic intervals and fixed getIntervalsByIds.

sruffell commented:
After updating most of the tests in this pull request, I realized that when I first added support for the modify command I did not handle synthetic intervals.  This change really isn't specifically related to performance improvements, but it just seemed easier to add it to this series since the easiest fix uses the `flattenDatabase` helper added as part of it.

lauft commented on src/data.cpp:
```diff
@@ -625,7 +731,7 @@ std::vector <Interval> getTracked (
   // Assign an ID to each interval.
   for (unsigned int i = 0; i < intervals.size (); ++i)
-    intervals[i].id = intervals.size () - i;
+    intervals[i].id = intervals.size () - i + id_skip;
```
Please use curly braces for `for` and `if` statements for new and modified code 😉

sruffell commented on test/move.t:
```diff
@@ -133,25 +133,70 @@ class TestMove(TestCase):
+                                expectedTags=[],
+                                description="unmodified interval")
+
+        # First move the non-synthetic one
```
> I do not fully understand why you needed to add the non-synthetic interval here...
> 
> Do you want to test the case where Timewarrior moves a non-synthetic interval into an exclusion?
> But then the interval from `move.t:146`
> 
> * is not moved in to an exclusion (these are defined for all days hours before the interval), and
> * there are already tests for moving closed intervals, e.g. `move.t:46`

I had an error in getIntervalsById in an early iteration of the pull request which the tests did not catch.

When there was a mix of synthetic and non-synthetic intervals, the open interval at the end of the database was counted twice. Below is the difference between that early iteration and what you eventually merged:

```diff
diff --git a/src/data.cpp b/src/data.cpp
index f7395e9..fab33d1 100644
--- a/src/data.cpp
+++ b/src/data.cpp
@@ -344,13 +344,25 @@ std::vector <Interval> getIntervalsByIds (
   auto id_it = ids.begin ();
   auto id_end = ids.end ();

+  auto it = database.begin ();
+  auto end = database.end ();
+
+  Interval latest;
+  if (it != end )
+  {
+    latest = IntervalFactory::fromSerialization (*it);
+  }
+
   // If the latest interval is open, check for synthetic intervals
-  auto latest = getLatestInterval (database);
   if (latest.is_open ())
   {
     auto exclusions = getAllExclusions (rules, {latest.start, Datetime ()});
     if (! exclusions.empty ())
     {
+        // We're converting the latest interval into synthetic intervals so we
+        // need to skip it
+        ++it;
+
         for (auto& interval : flatten (latest, exclusions))
         {
           ++current_id;
@@ -379,7 +391,7 @@ std::vector <Interval> getIntervalsByIds (
   current_id = synthetic.size ();

   // We'll find remaining intervals from the database itself
-  for (auto& line : database)
+  for ( ; it != end; ++it)
   {
     ++current_id;

@@ -388,7 +400,7 @@ std::vector <Interval> getIntervalsByIds (

     if (current_id == *id_it)
     {
-      Interval interval = IntervalFactory::fromSerialization (line);
+      Interval interval = IntervalFactory::fromSerialization (*it);
       interval.id = current_id;
       intervals.push_back (interval);
       ++id_it;
```

So I edited some of the existing tests just to add this non-synthetic interval to make sure this specific state of the database would be caught in the future.

> Generally I would prefer to limit tests on one aspect only and rather have separate tests then.

I was being a little lazy ;)  I'll restore the original tests and add new ones for explicitly testing this state in the follow-up pull request.

sruffell commented on test/modify.t:
```diff
@@ -166,6 +166,85 @@ class TestModify(TestCase):
         code, out, err = self.t.runError("modify start @1 {:%Y-%m-%dT%H:%M:%S}Z".format(now_utc - timedelta(hours=2)))
         self.assertIn("You cannot overlap intervals", err)
 
+    def test_modify_move_start_of_synthetic_interval(self):
```
The pull request will have this test split into two that both contain a mixed of synthetic and non-synthetic intervals.
